### DISCUSSION TITLE
Create usage tests in the CI

### DISCUSSION THREAD

**Discussion by talregev** [microsoft/vcpkg#40725] [Ideas]
Create usage tests in the CI
@dg0yt @autoantwort 
What do you think? How we can implement such a thing?

**Comment by autoantwort** [comment]
I have an idea how to implement that, but I don't think that the vcpkg team has time to review and merge it...

**Reply by talregev** [reply]
> I have an idea how to implement that, but I don't think that the vcpkg team has time to review and merge it...

Maybe they will invest the extra time review it, because it will save them a lot of time in review.

**Reply by autoantwort** [reply]
There is also https://github.com/microsoft/vcpkg-tool/pull/802 which would same them a lot of time in review and no response since a year ...

**Comment by talregev** [comment]
@BillyONeal What do you think?

**Comment by dg0yt** [comment]
For now, the solution is: **test ports**.
AFAICT it scales well, it is accepted without discussions, and with very little code, you can test more than the vcpkg crew would be able to test manually: You choose which headers to include, which symbols to use, and which extra cmake code to call.
And for very important ports, you list them with `=pass` in the CI baseline.

Slightly extended example/template: 
https://github.com/microsoft/vcpkg/tree/master/scripts/test_ports/vcpkg-ci-soci

**Reply by talregev** [reply]
> For now, the solution is: **test ports**. AFAICT it scales well, it is accepted without discussions, and with very little code, you can test more than the vcpkg crew would be able to test manually: You choose which headers to include, which symbols to use, and which extra cmake code to call. And for very important ports, you list them with `=pass` in the CI baseline.
> 
> Slightly extended example/template: https://github.com/microsoft/vcpkg/tree/master/scripts/test_ports/vcpkg-ci-soci

The manual solution is good. But it not scale. Most of the people don't do this tests, and sometimes port when they check, they don't check usage of every os they support by lack of time.

That why we need some automation in ci that scale up and save time.

**Reply by dg0yt** [reply]
THESE TESTS DO RUN IN CI.

**Reply by talregev** [reply]
I understand that these tests run in the CI, but you need manually to write them down for each port that you add / each os that you support. It not scale up. most of the ports don't have these tests. Do you see that?

**Reply by Osyotr** [reply]
> I understand that these tests run in the CI, but you need manually to write them down for each port that you add / each os that you support. It not scale up. most of the ports don't have these tests. Do you see that?

Between "most ports don't have tests" and "most ports have tests" what do you think should happen?

**Reply by talregev** [reply]
As I say in the description, I think we should build automatic tests that for every port that we have, it will test the usage. 
Then All the port will have a usage tests. 

As your comment imply that we should add them manually it will not scale up, and no one do it. That why I open this discussion.
And as @autoantwort comment, he have an idea to how to do it.

**Reply by dg0yt** [reply]
Writing them manually isn't too difficult (in contrast to writing or updating ports), and a one-time effort per port which needs it. A lot of stuff is also (somewhat) covered by transitive dependencies of a few monster ports.

FTR we already submitted PRs to run simple usage checks as part of the portfile. But it is not desired to run this every time a port is built. Only in CI or on demand.

**Reply by talregev** [reply]
> already submitted PRs to run simple usage checks as part of the portfile.

This test is not all on os, and sometimes we miss things like happen in my PRs. 
Also I need to wait for someone will test my port then to understand the problem, instead the CI that always available to me. 
I think the stuff in vcpkg have a lot of work to do, and they cannot allocate time to add a new test to every new port that added.

**Reply by dg0yt** [reply]
> This test is not all on os, 

You must be talking about different things.

> Also I need to wait for someone will test my port

Because you don't spend 60 seconds on a test port which would pay back quickly.

>  I think the stuff in vcpkg have a lot of work to do, and they cannot allocate time to add a new test to every new port that added.

Their work is full of surprises. Obviously, they write test projects. Everytime. Again. Even they would save time if they would do it only once as an added test port.

**Reply by talregev** [reply]
> You must be talking about different things.
> 
> >

We are talking the same thing. For example on the ls-qpack they test the usage only on windows, but there was a problem the usage on linux: https://github.com/microsoft/vcpkg/pull/40710



> Because you don't spend 60 seconds on a test port which would pay back quickly.
> 
> >

You don't spend the 60 seconds yourself when you update ports.



> Their work is full of surprises. Obviously, they write test projects. Everytime. Again. Even they would save time if they would do it only once as an added test port.

You can write them and ask them to change the policy, but I don't think is the solution. I think automatic tests is the solution.

**Reply by dg0yt** [reply]
> For example on the ls-qpack they test the usage only on windows, but there was a problem the usage on linux: #40710

So you talk about manual tests. I talk about test ports which run in CI. Different things.

> You don't spend the 60 seconds yourself when you update ports.

No. Apart from pointed wording, this was explicitly about copying an existing test port and doing tiny adjustment to 4 files.

> You can write them and ask them to change the policy, but I don't think is the solution. I think automatic tests is the solution.

The problem is: The solution will not be available in 60 seconds, and probably not even in 60 weeks.
And the delay for some relevant tool PRs is actively discouraging contributions.

OTOH every second spend on writing a test port quickly pays back.

**Reply by dg0yt** [reply]
(Obsolete, but I will reduce my further input.)

**Reply by talregev** [reply]
> already submitted PRs to run simple usage checks as part of the portfile.

> So you talk about manual tests. I talk about test ports which run in CI. Different things.

We both talk about the manual tests when submitted a PR, and I continue to take about that.

And my main talk is about automatic tests.

**Reply by talregev** [reply]
Most of the people and most of the ports lake usage tests on the CI. People not do that even if 60 seconds and there is no policy that enforce that.

I think this is not the way, and we need to think about things that will scale up, meaning that it will not add more work for people when add or update ports.
